Library/ViewModels/SignupViewModel.swift

```diff
@@ -4,108 +4,85 @@ import Prelude
+    public let (emailTextChangedSignal, emailTextChangedObserver) = Signal<String?, NoError>.pipe()
+    /// Call when the user returns from email text field.
+    public let (emailTextFieldDidReturnSignal, emailTextFieldDidReturnObserver) = Signal<(), NoError>.pipe()
+    /// Call when the environment has been logged into
+    public let (environmentLoggedInSignal, environmentLoggedInObserver) = Signal<(), NoError>.pipe()
+    /// Call when the user enters a new name.
+    public let (nameTextChangedSignal, nameTextChangedObserver) = Signal<String?, NoError>.pipe()
+    /// Call when the user returns from the name text field.
+    public let (nameTextFieldDidReturnSignal, nameTextFieldDidReturnObserver) = Signal<(), NoError>.pipe()
+    /// Call when the user enters a new password.
+    public let (passwordTextChangedSignal, passwordTextChangedObserver) = Signal<String?, NoError>.pipe()
+    /// Call when the user returns from the password text field.
+    public let (passwordTextFieldDidReturnSignal, passwordTextFieldDidReturnObserver) =
+      Signal<(), NoError>.pipe()
+    /// Call when the user taps signup.
+    public let (signupButtonPressedSignal, signupButtonPressedObserver) = Signal<(), NoError>.pipe()
+    /// Call when the view did load.
+    public let (viewDidLoadSignal, viewDidLoadObserver) = Signal<(), NoError>.pipe()
+    /// Call when the user toggles weekly newsletter.
+    public let (weeklyNewsletterChangedSignal, weeklyNewsletterChangedObserver) = Signal<Bool, NoError>.pipe()
+
+    init() {}
+  }
 
-    public init() {
-    let initialText = self.viewDidLoadProperty.signal.mapConst("")
-    let name = Signal.merge(
-      self.nameChangedProperty.signal,
-      initialText
-    )
-    let email = Signal.merge(
-      self.emailChangedProperty.signal,
-      initialText
-    )
-    let password = Signal.merge(
-      self.passwordChangedProperty.signal,
-      initialText
-    )
+  public typealias Outputs = (
+    // Call when the emailTextField should become the first responder
+    emailTextFieldBecomeFirstResponder: Signal<(), NoError>,
+    // Call when the sign up button should be enabled
+    isSignupButtonEnabled: Signal<Bool, NoError>,
+    /// Emits an access token envelope that can be used to update the environment.
+    logIntoEnvironment: Signal<AccessTokenEnvelope, NoError>,
+    /// Sets whether the password text field is the first responder.
+    passwordTextFieldBecomeFirstResponder: Signal<(), NoError>,
+    /// Emits when a notification should be posted.
+    postNotification: Signal<Notification, NoError>,
+    /// Sets whether the name text field is the first responder.
+    nameTextFieldBecomeFirstResponder: Signal<(), NoError>,
+    /// Emits the value for the weekly newsletter.
+    setWeeklyNewsletterState: Signal<Bool, NoError>,
+    /// Emits when a signup error has occurred and a message should be displayed.
+    showError: Signal<String, NoError>
+  )
+
+  public let inputs = Inputs()
```

Comment:
Why do we have this local `let` again?

Reply:
We store a reference to the inputs in order to send them values (via self.vm.inputs.input.send(value: value)` in both the view controller and the view model tests, but also in order to pass inputs into the `outputs(for inputs:)` function to generate our outputs.

Reply:
Oh right... why do we need to pass it in as an argument if it's a local var? Couldn't `outputs()` take no arguments and return `self.inputs`?

Reply:
Yep, that could also work.

Reply:
Cool, yeah I think if there's no real reason to pass it in then that makes sense? It doesn't help us with tests as far as I can tell.